discovery-provider/plugins/pedalboard/src/app.ts

```diff
@@ -0,0 +1,159 @@
+import { Knex, knex } from "knex"
+import { setIntervalAsync } from "set-interval-async"
+import { Table } from "./models"
+
+export default class App<AppData> {
+    // database connections
+    private dnDb: Knex
+    private idDb?: Knex
+
+    // pg notify handlers
+    private listeners: Map<string, ((self: App<AppData>, msg: any) => Promise<void>)[]>
+    // table scans
+    private scans: Map<string, ((self: App<AppData>, msg: any) => Promise<void>)[]>
+    // functions that execute on an interval
+    private repeaters: ([number, (self: App<AppData>) => Promise<void>])[]
+    // async operations that are evaluated immediately
+    private spawns: ((self: App<AppData>) => Promise<void>)[]
+
+    private appData: AppData
+
+    constructor(appData: AppData) {
+        this.dnDb = knex({

```

Comment:
nit, i would prefer to spell these out

this.discoveryNodeDatabase

just makes life easier to reason with I think, especially because we conflate discovery node & discovery provider